File: python/finder.py
"""
Module for finding shortest path between two points in graph
Time complexity: O(V*log(v)), where v - number of vertices
"""
import math
import argparse
from typing import Tuple, List
from binary_heap import BinaryHeap

Point = Tuple[int, int]


def find_shortest_path(heights: List[List[float]], step: float,
                       point1: Point,
                       point2: Point) -> Tuple[float, List[Tuple[int, int]]]:
    rows = len(heights)
    cols = len(heights[0])

    heap = BinaryHeap()
    heap.push((0., point1))

    MAX = 10000000000.0
    not_visited = ((-1, -1), MAX, False)
    parents = [[not_visited for _ in range(cols)] for _ in range(rows)]
    parents[point1[0]][point1[1]] = ((0, 0), 0., True)

    while heap:
        dist, [py, px] = heap.pop()
        if (py, px) == point2:
            break
        parents[py][px] = (parents[py][px][0], dist, True)
        for dx, dy in [(0, 1), (0, -1), (-1, 0), (1, 0)]:
            y = py + dy
            x = px + dx
            if 0 <= y < rows and 0 <= x < cols:
                height_diff = heights[py][px] - heights[y][x]
                new_dist = dist + math.sqrt(step ** 2 + height_diff ** 2)

                if not parents[y][x][2] and parents[y][x][1] > new_dist:
                    parents[y][x] = ((py, px), new_dist, False)
                    heap.push((new_dist, (y, x)))

    path = [point2]
    while path[-1] != point1:
        path.append(parents[path[-1][0]][path[-1][1]][0])
    return parents[path[0][0]][path[0][1]][1], path[::-1]


def run_test(path_to_test: str, dst: str):
    arr = []
    with open(path_to_test, 'r', encoding='utf-8') as f:
        lines = f.read().split('\n')
        start = tuple(map(int, lines[0].split(' ')))[::-1]
        end = tuple(map(int, lines[1].split(' ')))[::-1]
        step = float(lines[2])
        
        for line in lines[4:]:
            if line:
                arr.append(list(map(float, filter(len, line.split(' ')))))
        with open(dst, 'w', encoding='utf-8') as f:
            f.write(str(find_shortest_path(arr, step, start, end)))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("test", help="path to test")
    parser.add_argument("dst", help="dest path")
    args = parser.parse_args()
    run_test(args.test, args.dst)


if __name__ == "__main__":
    main()

File: cpp/finder.cpp
#include <bits/stdc++.h>
#include "defs.hpp"
#include "binary_heap.hpp"

using namespace std;

pair<double, vector<Vec2>> find_shortest_path(const vector<vector<double>> &heights, const Vec2 &start, const Vec2 &end, double step) {
    size_t rows = heights.size();
    size_t cols = heights[0].size();

    auto heap = BinaryHeap<Node>();
    heap.push({0., start});

    double MAX = 10000000000.0;
    Record not_visited = {{MAX, {-1, -1}}, false};
    vector<vector<Record>> parents(rows, vector<Record>(cols, not_visited));
    parents[start.y][start.x] = {{0, start}, true};

    while (!heap.empty()) {
        auto node = heap.pop();
        if (node.point == end)
            break;
        parents[node.point.y][node.point.x].visited = true;
        int neighbours[4][2] = {{0,  1},
                                {0,  -1},
                                {1,  0},
                                {-1, 0}};
        for (auto[dx, dy]: neighbours) {
            int x = node.point.x + dx;
            int y = node.point.y + dy;

            if (0 <= x && x < cols && 0 <= y && y < rows) {
                double height_diff = heights[node.point.y][node.point.x] - heights[y][x];
                double new_dist = node.distance + sqrt(step * step + height_diff * height_diff);

                if (!parents[y][x].visited && parents[y][x].node.distance > new_dist) {
                    parents[y][x] = {{new_dist, node.point}, false};
                    heap.push({new_dist, {x, y}});
                }
            }
        }
    }
    vector<Vec2> path = {end};
    while (path[path.size() - 1] != start)
        path.push_back(parents[path[path.size() - 1].y][path[path.size() - 1].x].node.point);
    auto dist = parents[path[0].y][path[0].x].node.distance;
    reverse(path.begin(), path.end());
    return {dist, path};
}

int main(int argc, char** argv) {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    freopen(argv[1], "r", stdin);
    Vec2 start, end;
    cin >> start.x >> start.y >> end.x >> end.y;
    int step;
    cin >> step;
    int rows, cols;
    cin >> rows >> cols;
    vector<vector<double>> heights(rows, vector<double>(cols));
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++)
            cin >> heights[i][j];
    }
    fclose(stdin);
    auto arr = find_shortest_path(heights, start, end, step);

    freopen(argv[2], "w", stdout);
    cout <<  "(" << fixed << setprecision(10) << arr.first << ", [";
    for (size_t i = 0; i < arr.second.size(); i++) {
        cout << "(" << arr.second[i].y << ", " << arr.second[i].x << ")";
        if (i != arr.second.size() - 1)
            cout << ", ";
    }
    cout << "])";
    cout.flush();
    fclose(stdout);
    return 0;
}
